compiler/noirc_evaluator/src/ssa/ir/instruction/call/blackbox.rs

```diff
@@ -70,52 +71,127 @@ pub(super) fn simplify_msm(
+
     match (dfg.get_array_constant(arguments[0]), dfg.get_array_constant(arguments[1])) {
         (Some((points, _)), Some((scalars, _))) => {
-            let Some(points) = points
-                .into_iter()
-                .map(|id| dfg.get_numeric_constant(id))
-                .collect::<Option<Vec<_>>>()
-            else {
-                return SimplifyResult::None;
-            };
-
-            let Some(scalars) = scalars
-                .into_iter()
-                .map(|id| dfg.get_numeric_constant(id))
-                .collect::<Option<Vec<_>>>()
-            else {
-                return SimplifyResult::None;
-            };
+            // We decompose points and scalars into constant and non-constant parts in order to simplify MSMs where a subset of the terms are constant.
+            let mut constant_points = vec![];
+            let mut constant_scalars_lo = vec![];
+            let mut constant_scalars_hi = vec![];
+            let mut var_points = vec![];
+            let mut var_scalars = vec![];
+            let len = scalars.len() / 2;
+            for i in 0..len {
+                match (
+                    dfg.get_numeric_constant(scalars[2 * i]),
+                    dfg.get_numeric_constant(scalars[2 * i + 1]),
+                    dfg.get_numeric_constant(points[3 * i]),
+                    dfg.get_numeric_constant(points[3 * i + 1]),
+                    dfg.get_numeric_constant(points[3 * i + 2]),
+                ) {
+                    (Some(lo), Some(hi), _, _, _)
+                        if lo == FieldElement::zero() && hi == FieldElement::zero() =>
+                    {
+                        is_constant = true;
+                        constant_scalars_lo.push(lo);
+                        constant_scalars_hi.push(hi);
+                        constant_points.push(FieldElement::zero());
+                        constant_points.push(FieldElement::zero());
+                        constant_points.push(FieldElement::one());
+                    }
+                    (_, _, _, _, Some(infinity)) if infinity == FieldElement::one() => {
+                        is_constant = true;
+                        constant_scalars_lo.push(FieldElement::zero());
+                        constant_scalars_hi.push(FieldElement::zero());
+                        constant_points.push(FieldElement::zero());
+                        constant_points.push(FieldElement::zero());
+                        constant_points.push(FieldElement::one());
+                    }
+                    (Some(lo), Some(hi), Some(x), Some(y), Some(infinity)) => {
+                        is_constant = true;
+                        constant_scalars_lo.push(lo);
+                        constant_scalars_hi.push(hi);
+                        constant_points.push(x);
+                        constant_points.push(y);
+                        constant_points.push(infinity);
+                    }
+                    _ => {
```

Comment:
Since this is run in a loop, won't it unset `is_constant` if _any_ of the values in the MSM are non-constant? I.e. currently appx. `is_constant = scalars.all(|scalar| scalar.is_constant())`?

I think setting `let mut is_constant = true;` at the top and then skipping this `is_constant = false` will have behavior closer to `scalars.any(|scalar| scalar.is_constant())`, which is what I'd expect from your comment:

>  We decompose points and scalars into constant and non-constant parts in order to simplify MSMs where a subset of the terms are constant.

Reply:
`is_constant` is properly set at each iteration because it is set in all branches of the match, so I think it is correct.